File: src/test/java/io/github/mjaroslav/bon2gradle/task/Bon2TaskTest.java
package io.github.mjaroslav.bon2gradle.task;

import io.github.mjaroslav.bon2gradle.test.shared.IOUtils;
import io.github.mjaroslav.bon2gradle.test.shared.JarComparison;
import lombok.val;
import org.gradle.testfixtures.ProjectBuilder;
import org.junit.jupiter.api.BeforeAll;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.io.TempDir;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;

import static io.github.mjaroslav.bon2gradle.test.shared.TestConstants.PACKAGE;

class Bon2TaskTest {
    @TempDir
    static File projectDirFile;

    static Path projectDir, deobfuscatedJar, deobfuscatedActualJar, obfuscatedJar, fieldsScv, methodsCsv;

    @BeforeAll
    static void beforeAll() {
        projectDir = projectDirFile.toPath();
        deobfuscatedJar = projectDir.resolve("Deobfuscated.jar");
        deobfuscatedActualJar = projectDir.resolve("DeobfuscatedActual.jar");
        obfuscatedJar = projectDir.resolve("Obfuscated.jar");
        methodsCsv = projectDir.resolve("methods.csv");
        fieldsScv = projectDir.resolve("fields.csv");
    }

    @Test
    void test$deobf() throws IOException {
        val project = ProjectBuilder.builder()
            .withProjectDir(projectDirFile)
            .build();
        val task = project.getTasks().create("deobf", Bon2Task.class);

        IOUtils.copyFromResources(PACKAGE + "task/Bon2TaskObfuscated.jar", deobfuscatedJar);
        IOUtils.copyFromResources(PACKAGE + "task/Bon2TaskDeobfuscated.jar", deobfuscatedActualJar);
        IOUtils.copyFromResources(PACKAGE + "task/Bon2TaskMethods.csv", methodsCsv);
        IOUtils.copyFromResources(PACKAGE + "task/Bon2TaskFields.csv", fieldsScv);

        task.getFrom().set(obfuscatedJar.toFile());
        task.getTo().set(deobfuscatedActualJar.toFile());
        task.getMapping().set(projectDir.toFile());
        task.getMappingsRelativeConfPath().set(".");
        task.doTask();

        JarComparison.compareJarClassMembers(deobfuscatedJar, deobfuscatedActualJar);
    }
}

File: src/test/kotlin/Bon2TaskTest.kt
package io.github.mjaroslav.bon2gradle.task

import io.github.mjaroslav.bon2gradle.test.shared.PACKAGE
import io.github.mjaroslav.bon2gradle.test.shared.compareJarClassMembers
import io.github.mjaroslav.bon2gradle.test.shared.copyFromResources
import org.gradle.testfixtures.ProjectBuilder
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.TestInstance
import org.junit.jupiter.api.io.TempDir
import java.nio.file.Path

@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class Bon2TaskTest {
    lateinit var projectDir: Path
    lateinit var deobfuscatedJar: Path
    lateinit var deobfuscatedActualJar: Path
    lateinit var obfuscatedJar: Path
    lateinit var fieldsScv: Path
    lateinit var methodsCsv: Path

    @BeforeAll
    fun beforeAll(@TempDir projectDir: Path) {
        this.projectDir = projectDir
        deobfuscatedJar = projectDir.resolve("Deobfuscated.jar")
        deobfuscatedActualJar = projectDir.resolve("DeobfuscatedActual.jar")
        obfuscatedJar = projectDir.resolve("Obfuscated.jar")
        methodsCsv = projectDir.resolve("methods.csv")
        fieldsScv = projectDir.resolve("fields.csv")
    }

    @Test
    fun `test$deobf`() {
        val project = ProjectBuilder.builder()
            .withProjectDir(projectDir.toFile())
            .build()
        val task = project.tasks.create("deobf", Bon2Task::class.java)
        copyFromResources("${PACKAGE}task/Bon2TaskObfuscated.jar", obfuscatedJar)
        copyFromResources("${PACKAGE}task/Bon2TaskDeobfuscated.jar", deobfuscatedJar)
        copyFromResources("${PACKAGE}task/Bon2TaskMethods.csv", methodsCsv)
        copyFromResources("${PACKAGE}task/Bon2TaskFields.csv", fieldsScv)
        task.from.set(obfuscatedJar.toFile())
        task.to.set(deobfuscatedActualJar.toFile())
        task.mapping.set(projectDir.toFile())
        task.mappingRelativeConfPath.set(".")
        task.doTask()
        compareJarClassMembers(deobfuscatedJar, deobfuscatedActualJar)
    }
}
